File: test/_harnesses/zig/src/packed_broken.zig
const std = @import("std");
const Checker = @import("./util.zig").Checker;

const small = @import("lib/SmallMessages.zig");

const msg_list = [_]small.Message{
    small.Message{ .IntMessage = small.IntMessage{} },
    small.Message{ .FloatMessage = small.FloatMessage{} },
    small.Message{ .FloatMessage = small.FloatMessage{} },
    small.Message{ .FloatMessage = small.FloatMessage{} },
    small.Message{ .IntMessage = small.IntMessage{} },
    small.Message{ .EmptyMessage = small.EmptyMessage{} },
    small.Message{ .LongMessage = small.LongMessage{} },
    small.Message{ .LongMessage = small.LongMessage{} },
    small.Message{ .LongMessage = small.LongMessage{} },
    small.Message{ .IntMessage = small.IntMessage{} },
};

pub fn main() !void {
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const testAllocator = gpa.allocator();

    var checker = Checker{};

    if (std.mem.eql(u8, args[1], "--generate")) {
        const buffer_size = small.getPackedSize(&msg_list);
        const buffer = try testAllocator.alloc(u8, buffer_size);
        defer testAllocator.free(buffer);

        small.packMessages(&msg_list, buffer);

        buffer[4] = 15;

        var file = try std.fs.cwd().createFile(args[2], .{ .truncate = true });
        defer file.close();

        try file.writeAll(buffer);
    } else if (std.mem.eql(u8, args[1], "--read")) {
        var file = try std.fs.cwd().openFile(args[2], .{});
        defer file.close();

        const buffer = try file.readToEndAlloc(testAllocator, std.math.maxInt(u32));
        defer testAllocator.free(buffer);

        if (small.unpackMessages(testAllocator, buffer)) |_| {
            checker.softAssert(false, "broken unpack");
        } else |err| {
            checker.softAssert(err == error.InvalidData, "broken unpack error");
        }
    }

    const mem_result = gpa.deinit();
    checker.softAssert(mem_result == .ok, "memory leaks");

    checker.check();
}

File: test/_harnesses/rust/src/packed_broken.rs
// message files will include code that's not necessarily called in each test
#![allow(dead_code)]

use std::fs;

mod small_messages;
use small_messages::*;

mod util;

fn main() {
    let mut checker = util::Checker { ok: true };
    let args = util::arg_parse();

    let msg_list = vec![
        Message::IntMessage(IntMessage::default()),
        Message::FloatMessage(FloatMessage::default()),
        Message::FloatMessage(FloatMessage::default()),
        Message::FloatMessage(FloatMessage::default()),
        Message::IntMessage(IntMessage::default()),
        Message::EmptyMessage(EmptyMessage::default()),
        Message::LongMessage(LongMessage::default()),
        Message::LongMessage(LongMessage::default()),
        Message::LongMessage(LongMessage::default()),
        Message::IntMessage(IntMessage::default()),
    ];

    if args.contains_key("generate") {
        let mut writer: Vec<u8> = Vec::new();

        pack_messages(&msg_list, &mut writer);
        writer[4] = 15;

        let filename = args.get("generate").unwrap();
        fs::write(filename, writer).unwrap();
    }
    else if args.contains_key("read") {
        let filename = args.get("read").unwrap();
        let buffer = fs::read(&filename).unwrap();
        let mut reader = BufferReader::from_vec(buffer);
        match unpack_messages(&mut reader) {
            Ok(_) => checker.soft_assert(false, "broken unpack"),
            Err(e) => match e {
                small_messages::SmallMessagesError::InvalidData => {},
                _ => checker.soft_assert(false, "broken unpack error")
            }
        };
    }

    checker.check();
}
